ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss_backward/device/moreh_nll_loss_backward_device_operation.hpp

```diff
@@ -0,0 +1,107 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <optional>
+#include <variant>
+
+#include "ttnn/decorators.hpp"
+#include "ttnn/device_operation.hpp"
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/types.hpp"
+
+namespace ttnn::operations::moreh::moreh_nll_loss_backward {
+
+struct MorehNllLossBackwardDeviceOperation {
+    // Define the operation attributes. This is it to store all variables needed by operations that aren't tensors
+    struct operation_attributes_t {
+        const bool reduction_mean;
+        const int32_t ignore_index;
+        const MemoryConfig memory_config;
+        std::optional<const DeviceComputeKernelConfig> compute_kernel_config;
+    };
+
+    // Define the tensor arguments. This is it to store all tensors passed in and/or out of the operation
+    // Tensor arguments don't need to be just input tensors, they can be output tensors, input/output tensors, optional
+    // tensors, etc.
+    struct tensor_args_t {
+        const Tensor& target_tensor;
+        const Tensor& output_grad_tensor;
+        const std::optional<const Tensor> weight_tensor;
+        const std::optional<const Tensor> divisor_tensor;
+        const std::optional<const Tensor> input_grad_tensor;
+    };
+
+    // Define the return types for the shape(s) of the operation
+    // Can be a single ttnn::Shape, std::optional<ttnn::Shape>, std::vector<ttnn::Shape>, std::tuple<ttnn::Shape> etc.
+    using shape_return_value_t = ttnn::Shape;
+
+    // Define the return types for the tensor(s) of the operation
+    // Can be a single Tensor, std::optional<Tensor, ...>, std::vector<Tensor>, std::tuple<Tensor, ...> etc.
+    using tensor_return_value_t = ttnn::Tensor;
+
+    struct Factory {
+        // Shared variables are the variables that are shared between the create and override_runtime_arguments methods
+        struct shared_variables_t {
+            KernelHandle unary_reader_kernel_id;
+            KernelHandle unary_writer_kernel_id;
+            std::size_t num_cores;
+            std::size_t num_cores_y;
+        };
+        using cached_program_t = ttnn::device_operation::CachedProgram<shared_variables_t>;
+
+        static cached_program_t create(
+            const operation_attributes_t& operation_attributes,
+            const tensor_args_t& tensor_args,
+            tensor_return_value_t& tensor_return_value);
+
+        static void override_runtime_arguments(
+            cached_program_t& cached_program,
+            const operation_attributes_t& operation_attributes,
+            const tensor_args_t& tensor_args,
+            tensor_return_value_t& tensor_return_value);
+    };
+
+    using program_factory_t = std::variant<Factory>;
```

Comment:
Since the code for factory is too long, how about splitting it into sets for 2D, 3D, and 4D?